Make elasticsearch/cluster_stats metricset work for Stack Monitoring without xpack.enabled flag

sayden opened the pull request:
Ready to test in Kibana. data.json is included to help troubleshooting and it has been generating testing the Metricbeat binary directly with Elasticsearch

elasticmachine commented:
Pinging @elastic/stack-monitoring (Stack monitoring)

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21099/detail/PR-21099/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21099/detail/PR-21099/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21099/detail/PR-21099/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21099/detail/PR-21099/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_21099.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #21099 event]`

* Start Time: 2020-09-15T15:25:38.596+0000

* Duration: 14 min 4 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Make check`
    * Description: make  check
    
    * Duration: 6 min 28 sec
    
    * Start Time: 2020-09-15T15:32:14.657+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-21099/runs/1/steps/339/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-09-15T15:38:53.358Z] Failed in branch Elastic Agent x-pack Windows
[2020-09-15T15:38:53.369Z] Failed in branch Elastic Agent Mac OS X
[2020-09-15T15:38:53.378Z] Failed in branch Filebeat oss
[2020-09-15T15:38:53.389Z] Failed in branch Filebeat x-pack
[2020-09-15T15:38:53.400Z] Failed in branch Filebeat Mac OS X
[2020-09-15T15:38:53.412Z] Failed in branch Filebeat x-pack Mac OS X
[2020-09-15T15:38:53.422Z] Failed in branch Filebeat Windows
[2020-09-15T15:38:53.432Z] Failed in branch Filebeat x-pack Windows
[2020-09-15T15:38:53.441Z] Failed in branch Heartbeat oss
[2020-09-15T15:38:53.451Z] Failed in branch Heartbeat Mac OS X
[2020-09-15T15:38:53.461Z] Failed in branch Heartbeat Windows
[2020-09-15T15:38:53.470Z] Failed in branch Auditbeat oss Linux
[2020-09-15T15:38:53.480Z] Failed in branch Auditbeat crosscompile
[2020-09-15T15:38:53.490Z] Failed in branch Auditbeat oss Mac OS X
[2020-09-15T15:38:53.500Z] Failed in branch Auditbeat oss Windows
[2020-09-15T15:38:53.510Z] Failed in branch Auditbeat x-pack
[2020-09-15T15:38:53.521Z] Failed in branch Auditbeat x-pack Mac OS X
[2020-09-15T15:38:53.532Z] Failed in branch Auditbeat x-pack Windows
[2020-09-15T15:38:53.543Z] Failed in branch Libbeat x-pack
[2020-09-15T15:38:53.553Z] Failed in branch Metricbeat OSS Unit tests
[2020-09-15T15:38:53.563Z] Failed in branch Metricbeat OSS Go Integration tests
[2020-09-15T15:38:53.573Z] Failed in branch Metricbeat OSS Python Integration tests
[2020-09-15T15:38:53.583Z] Failed in branch Metricbeat crosscompile
[2020-09-15T15:38:53.593Z] Failed in branch Metricbeat Mac OS X
[2020-09-15T15:38:53.604Z] Failed in branch Metricbeat x-pack Mac OS X
[2020-09-15T15:38:53.614Z] Failed in branch Metricbeat Windows
[2020-09-15T15:38:53.624Z] Failed in branch Metricbeat x-pack Windows
[2020-09-15T15:38:53.634Z] Failed in branch Packetbeat Linux
[2020-09-15T15:38:53.644Z] Failed in branch Packetbeat Mac OS X
[2020-09-15T15:38:53.655Z] Failed in branch Packetbeat Windows
[2020-09-15T15:38:53.666Z] Failed in branch dockerlogbeat
[2020-09-15T15:38:53.676Z] Failed in branch Winlogbeat oss
[2020-09-15T15:38:53.686Z] Failed in branch Winlogbeat Windows
[2020-09-15T15:38:53.696Z] Failed in branch Winlogbeat Windows x-pack
[2020-09-15T15:38:53.706Z] Failed in branch Functionbeat x-pack
[2020-09-15T15:38:53.731Z] Failed in branch Functionbeat Mac OS X x-pack
[2020-09-15T15:38:53.741Z] Failed in branch Functionbeat Windows
[2020-09-15T15:38:53.751Z] Failed in branch Journalbeat
[2020-09-15T15:38:53.760Z] Failed in branch Generators Metricbeat Mac OS X
[2020-09-15T15:38:53.769Z] Failed in branch Generators Beat Mac OS X
[2020-09-15T15:38:53.779Z] Failed in branch Kubernetes
[2020-09-15T15:38:54.026Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-09-15T15:38:54.037Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-09-15T15:38:54.048Z] Stage "Generators" skipped due to earlier failure(s)
[2020-09-15T15:38:54.792Z] Failed in branch Metricbeat x-pack
[2020-09-15T15:38:54.802Z] Failed in branch Generators
[2020-09-15T15:38:54.853Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-09-15T15:38:55.048Z] Failed in branch Libbeat
[2020-09-15T15:39:02.707Z] [INFO] unstashV2: JOB_GCS_BUCKET is set. bucket param got precedency instead.
[2020-09-15T15:39:02.751Z] [INFO] unstashV2: JOB_GCS_CREDENTIALS is set. credentialsId param got precedency instead.
[2020-09-15T15:39:02.831Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs://beats-ci-temp/Beats/beats/PR-21099-1/source/source.tgz
[2020-09-15T15:39:02.851Z] [Google Cloud Storage Plugin] Downloading: Beats/beats/PR-21099-1/source/source.tgz to local path: /var/lib/jenkins/workspace/Beats_beats_PR-21099/source.tgz
[2020-09-15T15:39:11.281Z] + tar --version
[2020-09-15T15:39:11.601Z] + tar -xpf source.tgz
[2020-09-15T15:39:21.969Z] + rm source.tgz
[2020-09-15T15:39:21.997Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-21099/src/github.com/elastic/beats
[2020-09-15T15:39:22.047Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-21099/src/github.com/elastic/beats/Lint
[2020-09-15T15:39:22.577Z] + cat
[2020-09-15T15:39:22.577Z] + /usr/local/bin/runbld ./runbld-script --job-name elastic+beats+pull-request
[2020-09-15T15:39:22.577Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-09-15T15:39:29.163Z] runbld>>> runbld started
[2020-09-15T15:39:29.163Z] runbld>>> 1.6.12/f45d832f2ba0aa2722ab4ec1fda8ad140f027f8b
[2020-09-15T15:39:30.547Z] runbld>>> The following profiles matched the job 'elastic+beats+pull-request' in order of occurrence in the config (last value wins).
[2020-09-15T15:39:30.547Z] runbld>>> Matches in the system config:
[2020-09-15T15:39:30.547Z] runbld>>> - Matched ^elastic\+beats
[2020-09-15T15:39:30.547Z] runbld>>> - Matched ^elastic\+beats\+pull-request
[2020-09-15T15:39:31.932Z] runbld>>> Debug logging enabled.
[2020-09-15T15:39:31.932Z] runbld>>> Storing result
[2020-09-15T15:39:32.192Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-09-15T15:39:32.192Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200915153931-E51D1805
[2020-09-15T15:39:32.192Z] runbld>>> Adding system facts.
[2020-09-15T15:39:33.137Z] runbld>>> Adding vcs info for the latest commit:  2b98e83e8e62d66b72d225128f25f88308da3260
[2020-09-15T15:39:33.137Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-09-15T15:39:33.137Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-09-15T15:39:33.137Z] Processing JUnit reports with runbld...
[2020-09-15T15:39:33.137Z] + echo 'Processing JUnit reports with runbld...'
[2020-09-15T15:39:33.708Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-09-15T15:39:33.708Z] runbld>>> DURATION: 19ms
[2020-09-15T15:39:33.708Z] runbld>>> STDOUT: 40 bytes
[2020-09-15T15:39:33.708Z] runbld>>> STDERR: 49 bytes
[2020-09-15T15:39:33.708Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-09-15T15:39:33.708Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats_PR-21099
[2020-09-15T15:39:34.652Z] runbld>>> Storing build metadata: 
[2020-09-15T15:39:34.652Z] runbld>>> Adding test report.
[2020-09-15T15:39:34.652Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats_PR-21099/src/github.com/elastic/beats
[2020-09-15T15:39:35.223Z] runbld>>> Found 0 test output files
[2020-09-15T15:39:35.223Z] runbld>>> Test output logs contained: Errors: 0 Failures: 0 Tests: 0 Skipped: 0
[2020-09-15T15:39:35.484Z] runbld>>> Storing result
[2020-09-15T15:39:35.745Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-09-15T15:39:35.745Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200915153931-E51D1805
[2020-09-15T15:39:35.745Z] runbld>>> Email notification disabled by environment variable.
[2020-09-15T15:39:35.745Z] runbld>>> Slack notification disabled by environment variable.
[2020-09-15T15:39:41.520Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-21099
[2020-09-15T15:39:41.866Z] [INFO] getVaultSecret: Getting secrets
[2020-09-15T15:39:41.962Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-09-15T15:39:43.148Z] + chmod 755 generate-build-data.sh
[2020-09-15T15:39:43.148Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21099/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21099/runs/1 FAILURE 844277
[2020-09-15T15:39:43.148Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21099/runs/1/steps/?limit=10000 -o steps-info.json
[2020-09-15T15:39:43.699Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21099/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-09-15T15:39:43.699Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21099/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

chrisronline commented:
I'm seeing this while running this PR:

```
2020-09-16T14:19:17.647-0400	ERROR	runtime/map_faststr.go:204	recovered from panic while fetching 'elasticsearch/cluster_stats' for host 'localhost:9200'. Recovering, but please report this.	{"panic": "assignment to entry in nil map", "stack": "github.com/elastic/beats/v7/libbeat/logp.Recover\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/libbeat/logp/global.go:101\nruntime.gopanic\n\t/usr/local/Cellar/go/1.13.7/libexec/src/runtime/panic.go:679\nruntime.mapassign_faststr\n\t/usr/local/Cellar/go/1.13.7/libexec/src/runtime/map_faststr.go:204\ngithub.com/elastic/beats/v7/libbeat/common.MapStr.Put\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/libbeat/common/mapstr.go:199\ngithub.com/elastic/beats/v7/metricbeat/module/elasticsearch/cluster_stats.eventMapping\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:229\ngithub.com/elastic/beats/v7/metricbeat/module/elasticsearch/cluster_stats.(*MetricSet).Fetch\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/module/elasticsearch/cluster_stats/cluster_stats.go:70\ngithub.com/elastic/beats/v7/metricbeat/mb/module.(*metricSetWrapper).fetch\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:256\ngithub.com/elastic/beats/v7/metricbeat/mb/module.(*metricSetWrapper).startPeriodicFetching\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:224\ngithub.com/elastic/beats/v7/metricbeat/mb/module.(*metricSetWrapper).run\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:208\ngithub.com/elastic/beats/v7/metricbeat/mb/module.(*Wrapper).Start.func1\n\t/Users/chris/dev/repos/go/src/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:147"}
```

chrisronline commented:
I'm seeing this error now:
```
{"type":"mapper_parsing_exception","reason":"failed to parse field [elasticsearch.cluster.stats.nodes.total] of type [long] in document with id 'CaJUwHQBA9bBTy50HQpt'. Preview of field's value: '{total=1}'","caused_by":{"type":"json_parse_exception","reason":"Current token (START_OBJECT) not numeric, can not use numeric value accessors\n at [Source: (byte[])\"{\"@timestamp\":\"2020-09-24T13:36:11.140Z\",\"elasticsearch\":{\"cluster\":{\"name\":\"elasticsearch\",\"id\":\"PChw572_S52efCun1M0eNg\",\"stats\":{\"indices\":{\"store\":{\"size\":{\"bytes\":1211431477},\"reserved\":{\"bytes\":0}},\"query_cache\":{\"hit\":{\"total\":0},\"miss\":{\"total\":26},\"cache\":{\"total\":0},\"evictions\":0,\"cache_size\":{\"bytes\":0},\"memory_size\":{\"bytes\":0},\"total\":26},\"segments\":{\"total\":396,\"memory\":{\"bytes\":5437784,\"doc_values\":{\"bytes\":3400816},\"fixed_bit_set\":{\"bytes\":316224},\"points\":{\"bytes\":0},\"terms\":{\"ve\"[truncated 7233 bytes]; line: 1, column: 6281]"}}
```

Are you able to run this successfully against ES?

chrisronline commented:
I'm seeing this error on startup still:

```
2020-10-02T15:31:19.433-0400	WARN	[elasticsearch]	elasticsearch/client.go:407	Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Time{wall:0xbfd5fde073c27530, ext:48941485, loc:(*time.Location)(0x8416f00)}, Meta:null, Fields:{"agent":{"ephemeral_id":"7971d271-a0a5-4774-8e5c-ec770c1db6e5","id":"5ca7fc45-dad9-4ed5-afa0-ff17fd83a2e1","name":"cbr-mbp.lan","type":"metricbeat","version":"8.0.0"},"ecs":{"version":"1.5.0"},"elasticsearch":{"cluster":{"id":"OQ5paLC_Rn2qEVfCyRVg6Q","name":"elasticsearch","stats":{"cluster_name":"elasticsearch","indices":{"docs":{"deleted":{"total":0},"total":7},"fielddata":{"memory":{"bytes":0}},"query_cache":{"cache":{"total":0.000000},"cache_size":{"bytes":0},"evictions":0,"hit":{"total":0.000000},"memory_size":{"bytes":0},"miss":{"total":0.000000},"total":0},"segments":{"max_unsafe_auto_id":{"ms":-1},"memory":{"bytes":6500,"doc_values":{"bytes":380},"fixed_bit_set":{"bytes":0},"index_writer":{"bytes":0},"norms":{"bytes":0},"points":{"bytes":0},"stored_fields":{"bytes":2440},"terms":{"bytes":3680,"vectors":{"bytes":0}},"version_map":{"bytes":0}},"total":5},"shards":{"index":{"primaries":{"avg":1.000000,"max":1.000000,"min":1.000000},"replication":{"avg":0.000000,"max":0.000000,"min":0.000000},"shards":{"avg":1.000000,"max":1.000000,"min":1.000000}},"primaries":1,"replication":0,"total":1},"store":{"reserved":{"bytes":0},"size":{"bytes":18686}},"total":1},"license":{"cluster_needs_tls":false,"issue_date":"2020-10-02T19:30:16.399Z","issue_date_in_millis":1601667016399,"issued_to":"elasticsearch","issuer":"elasticsearch","max_nodes":1000,"start_date_in_millis":-1,"status":"active","type":"basic","uid":"50fbff47-0ccf-4eb0-8afb-e1df722163b6"},"nodes":{"coordinating_only":{"total":0},"data":{"total":1},"fs":{"available":{"bytes":222692048896},"free":{"bytes":239005847552},"total":{"bytes":499963174912}},"ingest":{"pipelines":{"total":1},"processor_stats":{"gsub":{"count":0.000000,"current":0.000000,"failed":0.000000,"time_in_millis":0.000000},"script":{"count":0.000000,"current":0.000000,"failed":0.000000,"time_in_millis":0.000000}},"total":1},"jvm":{"max_uptime":{"ms":71164},"memory":{"heap":{"max":{"bytes":1073741824},"used":{"bytes":101822472}}},"threads":{"total":68}},"master":{"total":1},"ml":{"total":1},"remote_cluster_client":{"total":1},"total":1,"transform":{"total":1},"voting_only":{"total":0}},"stack":{"apm":{"found":false},"xpack":{"ccr":{"auto_follow_patterns_count":0.000000,"available":false,"enabled":true,"follower_indices_count":0.000000},"data_streams":{"available":true,"data_streams":0.000000,"enabled":true,"indices_count":0.000000},"frozen_indices":{"available":true,"enabled":true,"indices_count":0.000000},"graph":{"available":false,"enabled":true},"ilm":{"policy":{"stats":[{"indices_managed":0.000000,"phases":{"delete":{"actions":["delete"],"min_age":7776000000.000000},"hot":{"actions":["rollover"],"min_age":0.000000}}},{"indices_managed":0.000000,"phases":{"hot":{"actions":["rollover"],"min_age":0.000000}}},{"indices_managed":0.000000,"phases":{"delete":{"actions":["delete"],"min_age":604800000.000000}}},{"indices_managed":0.000000,"phases":{"hot":{"actions":["rollover"],"min_age":0.000000}}},{"indices_managed":0.000000,"phases":{"hot":{"actions":["rollover"],"min_age":0.000000}}},{"indices_managed":0.000000,"phases":{"hot":{"actions":["rollover"],"min_age":0.000000}}},{"indices_managed":0.000000,"phases":{"delete":{"actions":["delete"],"min_age":7776000000.000000},"hot":{"actions":["rollover"],"min_age":0.000000}}}],"total":7}},"logstash":{"available":false,"enabled":true},"ml":{"available":false,"enabled":true,"jobs":{"total":0},"node":{"total":1}},"monitoring":{"available":true,"collection_enabled":false,"enabled":true,"enabled_exporters":{"local":1.000000}},"rollup":{"available":true,"enabled":true},"searchable_snapshots":{"available":false,"enabled":true,"indices_count":0.000000},"security":{"anonymous":{"enabled":false},"api_key_service":{"enabled":true},"audit":{"enabled":false},"available":true,"enabled":true,"fips_140":{"enabled":false},"ipfilter":{"http":false,"transport":false},"realms":{"active_directory":{"available":false,"enabled":false},"file":{"available":true,"cache":[{"size":0.000000}],"enabled":true,"name":["default_file"],"order":[-2147483647.000000],"size":[0.000000]},"kerberos":{"available":false,"enabled":false},"ldap":{"available":false,"enabled":false},"native":{"available":true,"cache":[{"size":0.000000}],"enabled":true,"name":["default_native"],"order":[-2147483646.000000],"size":[0.000000]},"oidc":{"available":false,"enabled":false},"pki":{"available":false,"enabled":false},"saml":{"available":false,"enabled":false}},"role_mapping":{"native":{"enabled":0.000000,"size":0.000000}},"roles":{"dls":{"bit_set_cache":{"count":0.000000,"memory":"0b","memory_in_bytes":0.000000}},"file":{"dls":false,"fls":false,"size":0.000000},"native":{"dls":false,"fls":false,"size":0.000000}},"ssl":{"http":{"enabled":true},"transport":{"enabled":false}},"token_service":{"enabled":true}},"slm":{"available":true,"enabled":true},"spatial":{"available":true,"enabled":true},"transform":{"available":true,"enabled":true},"vectors":{"available":true,"dense_vector_dims_avg_count":0.000000,"dense_vector_fields_count":0.000000,"enabled":true},"voting_only":{"available":true,"enabled":true}}},"state":{"cluster_uuid":"OQ5paLC_Rn2qEVfCyRVg6Q","master_node":"oqZqInokQGWbYdpwNwIEJA","nodes":{"oqZqInokQGWbYdpwNwIEJA":{"attributes":{"ml.machine_memory":"34359738368","ml.max_open_jobs":"20","transform.node":"true","xpack.installed":"true"},"ephemeral_id":"LPslIylaTwGjNCHms4g8DA","name":"cbr-mbp.lan","transport_address":"127.0.0.1:9300"}},"nodes_hash":-112262078,"state_uuid":"X6RT9q6fRluvMosidF4pdQ","status":"green","version":41.000000},"status":"green","version":"8.0.0-SNAPSHOT"}}},"event":{"dataset":"elasticsearch.cluster.stats","duration":336449522,"module":"elasticsearch"},"host":{"architecture":"x86_64","hostname":"cbr-mbp.lan","id":"0E799C52-1A23-58FB-8293-749EDE93385A","ip":["fe80::c2f:e735:1bac:b22d","192.168.86.227","fe80::aede:48ff:fe00:1122","fe80::d1:2983:3330:e64e","169.254.116.221","fe80::c135:20d0:94d9:dde2","fe80::fca8:b5:391d:542","fe80::2fe2:834:5714:9d4e","fe80::f36c:9877:dbc0:95fc"],"mac":["68:5b:35:cc:cb:1e","ac:de:48:00:11:22","3a:f9:d3:be:e5:4f","38:f9:d3:be:e5:4f","82:41:7e:c3:9c:01","82:41:7e:c3:9c:00","82:41:7e:c3:9c:05","82:41:7e:c3:9c:04","82:41:7e:c3:9c:01","0a:f9:d3:be:e5:4f","6e:a0:7b:dd:71:d9","6e:a0:7b:dd:71:d9"],"name":"cbr-mbp.lan","os":{"build":"19H2","family":"darwin","kernel":"19.6.0","name":"Mac OS X","platform":"darwin","version":"10.15.7"}},"metricset":{"name":"cluster_stats","period":10000},"service":{"address":"localhost:9200","type":"elasticsearch"}}, Private:interface {}(nil), TimeSeries:true}, Flags:0x0, Cache:publisher.EventCache{m:common.MapStr(nil)}} (status=400): {"type":"mapper_parsing_exception","reason":"failed to parse field [elasticsearch.cluster.stats.nodes.master] of type [long] in document with id 'kyDM6nQBVnm6CwgnHOYF'. Preview of field's value: '{total=1}'","caused_by":{"type":"json_parse_exception","reason":"Current token (START_OBJECT) not numeric, can not use numeric value accessors\n at [Source: (byte[])\"{\"@timestamp\":\"2020-10-02T19:31:13.868Z\",\"agent\":{\"version\":\"8.0.0\",\"ephemeral_id\":\"7971d271-a0a5-4774-8e5c-ec770c1db6e5\",\"id\":\"5ca7fc45-dad9-4ed5-afa0-ff17fd83a2e1\",\"name\":\"cbr-mbp.lan\",\"type\":\"metricbeat\"},\"metricset\":{\"name\":\"cluster_stats\",\"period\":10000},\"service\":{\"address\":\"localhost:9200\",\"type\":\"elasticsearch\"},\"elasticsearch\":{\"cluster\":{\"name\":\"elasticsearch\",\"id\":\"OQ5paLC_Rn2qEVfCyRVg6Q\",\"stats\":{\"nodes\":{\"jvm\":{\"threads\":{\"total\":68},\"max_uptime\":{\"ms\":71164},\"memory\":{\"heap\":{\"used\"[truncated 5663 bytes]; line: 1, column: 855]"}}
```

chrisronline commented:
@sayden 

I can now run this PR and it indexes properly! Thanks for fixing that!

I'm seeing an issue with the document though.

In existing `.monitoring-es-*` indices, we have:
```
POST .monitoring-es-*/_search?filter_path=hits.hits._source.cluster_stats.nodes.count
{
  "size": 1,
  "sort": [
    {
      "timestamp": {
        "order": "desc"
      }
    }
  ],
  "query": {
    "term": {
      "type": {
        "value": "cluster_stats"
      }
    }
  }
}
```

which returns:

```
{
  "hits" : {
    "hits" : [
      {
        "_source" : {
          "cluster_stats" : {
            "nodes" : {
              "count" : {
                "total" : 3,
                "coordinating_only" : 0,
                "data" : 2,
                "data_cold" : 0,
                "data_content" : 0,
                "data_hot" : 0,
                "data_warm" : 0,
                "ingest" : 2,
                "master" : 3,
                "ml" : 0,
                "remote_cluster_client" : 3,
                "transform" : 2,
                "voting_only" : 1
              }
            }
          }
        }
      }
    ]
  }
}

```

I'm not seeing that in the `metricbeat-*` documents. I see `elasticsearch.cluster.stats.nodes.total` as a number but I don't see an overall `count`. Is this by design? If so, that's fine and I can just adjust the code to look in either spot.

sayden commented:
The `total` vs `count` thing was my mistake, I have reverted all those fields to `count` again (which is the correct naming).

The error is more a misconception: You'll never find a `hits.hits._source.cluster_stats.nodes.count` now because that path doesn't exists in the stored JSON document. The only way to reach that info is to use `hits.hits._source.cluster.stats.nodes.count`. Example:
```
POST metricbeat-*/_search?filter_path=hits.hits._source.elasticsearch.cluster.stats.nodes.count
{
  "hits" : {
    "hits" : [
      {
        "_source" : {
          "elasticsearch" : {
            "cluster" : {
              "stats" : {
                "nodes" : {
                  "count" : 1
                }
              }
            }
          }
        }
      },
      {
        "_source" : {
          "elasticsearch" : {
            "cluster" : {
              "stats" : {
                "nodes" : {
                  "count" : 1
                }
              }
            }
          }
        }
      }
    ]
  }
}
```

Another solution is to create a mapping, which is included as an example in the PR now. But you'll have to "query" this info:
```
POST metricbeat-*/_search
{
  "_source": "elasticsearch.cluster.stats.nodes.count", 
  "query": {
    "term": {
      "cluster_stats.nodes.count": {
        "value": 1
      }
    }
  }
}

{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 17,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "metricbeat-8.0.0-2020.10.28-000001",
        "_id" : "vljIb3UBKtS08QdByHUG",
        "_score" : 1.0,
        "_source" : {
          "elasticsearch" : {
            "cluster" : {
              "stats" : {
                "nodes" : {
                  "count" : 1
                }
              }
            }
          }
        }
      }
    ]
  }
}
```

Note that we query the `term` `cluster_stats.nodes.count` but we request only the path in the source that is `elasticsearch.cluster.stats.nodes.count`, which is the real path in the doc.

I also removed some data from the incoming data from Elasticsearch to save some space. I have included it again just in case.

Sorry for the long explanation :sweat_smile:  I just wanted to make sure that I explained it properly.

sayden commented:
I forgot to push the commit that brings back the data at `nodes.count`

chrisronline commented:
I'm not seeing the `cluster_stats` documents anymore:

```
POST metricbeat-*/_search?filter_path=aggregations.types.buckets
{
  "size": 0,
  "aggs": {
    "types": {
      "terms": {
        "field": "metricset.name",
        "size": 10
      },
      "aggs": {
        "top": {
          "top_hits": {
            "size": 1,
            "sort": [
              {
                "@timestamp": "desc"
              }
            ],
            "_source": ""
          }
        }
      }
    }
  }
}
```

-> 

```
{
  "aggregations" : {
    "types" : {
      "buckets" : [
        {
          "key" : "node_stats",
          "doc_count" : 11,
          "top" : {
            "hits" : {
              "total" : {
                "value" : 11,
                "relation" : "eq"
              },
              "max_score" : null,
              "hits" : [
                {
                  "_index" : "metricbeat-8.0.0-2020.10.29-000001",
                  "_id" : "TRLLdHUBq5rHWs33_zDg",
                  "_score" : null,
                  "_source" : { },
                  "sort" : [
                    1603982326777
                  ]
                }
              ]
            }
          }
        }
      ]
    }
  }
}

```

chrisronline commented:
I'm seeing an error running this now:

> 2020-10-29T10:43:02.451-0400	ERROR	[publisher_pipeline_output]	pipeline/output.go:154	Failed to connect to backoff(elasticsearch(https://localhost:9200)): Connection marked as failed because the onConnect callback failed: error loading template: could not load template. Elasticsearch returned: couldn't load template: 400 Bad Request: {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]"}],"type":"mapper_parsing_exception","reason":"Failed to parse mapping: Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]","caused_by":{"type":"mapper_parsing_exception","reason":"Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]"}},"status":400}. Response body: {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]"}],"type":"mapper_parsing_exception","reason":"Failed to parse mapping: Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]","caused_by":{"type":"mapper_parsing_exception","reason":"Mapping definition for [count] has unsupported parameters:  [properties : {total={type=long}, data={type=long}, master={type=long}}]"}},"status":400}.

sayden commented:
Chris, can you briefly describe a step by step way to reproduce this error? Somehow it's happening in all metricsets but I cannot manage to reproduce it (I'm usually just guessing what the mapping issue is). The way I do it is:

1. Start a new Elasticsearch and Kibana container for each metricset (I remove the running container every time I start with a different metricset)
2. `go build` in the metricbeat folder
3. `./metricbeat modules disable system && .'/metricbeat modules enable elasticsearch`
4. Then, I configure the elasticsearch module (in the `modules.d` folder) to activate only the metricset being tested
5. I usually comment the processors in `metricbeat.yml` to remove some noise of the `add_*_metadata`
6. Finally I start metricbeat with `./metricbeat -e -d "*"`

From there, I don't get any error and I usually copy-paste the generated event from the line `processing/processors.go:187	Publish event: {` but I don't get any error.

chrisronline commented:
My steps which go against the same ES cluster

1. Pull down your PR
2. `cd metricbeat`
3. `mage build`
4. `./metricbeat modules enable elasticsearch-xpack` (same error with `elasticsearch` module too)
5. Delete all existing MB ES resources:
```
DELETE metricbeat-8.0.0* 
DELETE _template/metricbeat-8.0.0
```
6. `./metricbeat -c metricbeat.dev.yml -e`

Am I doing something wrong here?

sayden commented:
It must be me the one who is doing something wrong but I can't manage to understand why because, esentially, we are doing the same thing. You delete MB ES resources and I `docker rm -f elasticsearch` on each test. I'm wondering if it's reusing the same docker volume or something similar.

Anyways, the mapping error was fixed.

elasticmachine commented:
## :bug: Flaky test report
:grey_exclamation: There are test failures but not known flaky tests.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 7  |
| Passed       | 2146  |
| Skipped      | 123 |
| Total        | 2276   |



### Genuine test errors [![7](https://img.shields.io/badge/7%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21099/detail/PR-21099/9//tests)
:broken_heart: There are test failures but not known flaky tests, most likely a genuine test failure.

        * **Name**: `Build&Test / metricbeat-goIntegTest / TestFetch – elasticsearch`
        * **Name**: `Build&Test / metricbeat-goIntegTest / TestFetch/cluster_stats – elasticsearch`
        * **Name**: `Build&Test / metricbeat-goIntegTest / TestXPackEnabled – elasticsearch`
        * **Name**: `Build&Test / metricbeat-goIntegTest / TestXPackEnabled/cluster_stats – elasticsearch`
        * **Name**: `Build&Test / metricbeat-unitTest / [build failed] – cluster_stats`
        * **Name**: `Build&Test / metricbeat-pythonIntegTest / test_xpack_cluster_stats – metricbeat.module.elasticsearch.test_elasticsearch.Test`
        * **Name**: `Build&Test / metricbeat-windows-windows-2019 / [build failed] – cluster_stats`


</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_flaky.id-->

chrisronline reviewed (approved):
LGTM!

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/_meta/data.json:
```diff
@@ -1,43 +1,451 @@
-                        "count": 18,
-                        "primaries": 18
-                    },
-                    "total": 18
```
Looks like you changed this to `count`, which would be a breaking change. I would suggest reverting this to `total` again.

In general, we should not introduce breaking changes since this PR will be eventually backported to `7.x`. Its okay to introduce new fields in this PR but not okay to take away existing fields. To ensure this, I would suggest comparing the old `data.json` with the new `data.json` using something like jsondiff.com and making sure the changes from old => new are purely additive.

sayden replied:
Good catch! Thank you! I'll start using jsondiff, it will definitely help me checking backwards compatibility.

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/data.go:
```diff
@@ -19,29 +19,138 @@ package cluster_stats
+					"count": c.Int("ingest.number_of_pipelines"),
+				},
+				//TODO Not sure if to remove this one
+				"processor_stats": c.Ifc("ingest.processor_stats"),
```
Did you decide to keep this field? If so, remove the `TODO` comment so it's not confusing in the future. If not, what information are you looking for to make the decision? Is this field being used by the UI - if not, I think we can remove it then?

sayden replied:
Removed!

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/data.go:
```diff
@@ -50,30 +159,259 @@ var (
+					"stats": c.Ifc("policy_stats"),
+				},
+			}),
+			//"watcher": c.Ifc("watcher"),
```
Why is this field commented out?

sayden replied:
I forgot to add a comment for double check it's not required (I was hoping it wasn't). If Chris hasn't complain I guess it's not so I have removed it :smile:

chrisronline replied:
Is this data contained in the `stack_stats` field?

sayden replied:
It wasn't included when you first tested it, @chrisronline I guess it's safe to remove it

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/data_test.go:
```diff
@@ -21,10 +21,12 @@ package cluster_stats
 func TestMapper(t *testing.T) {
-	elasticsearch.TestMapperWithInfo(t, "./_meta/test/cluster_stats.*.json", eventMapping)
+	// TODO This test must be skipped until we find a reasonable way to test the event mapping when it has to do extra
+	// HTTP calls
```
Could you mock an HTTP server using [`http.NewServeMux`](https://golang.org/pkg/net/http/#NewServeMux)? IIRC we are doing this in a few places in the Beats codebase if you want to try and follow some examples.

sayden replied:
Ah yes, the problem I had with this test is the `helper.HTTP` which is not easy to use in tests. Actually it's not used in any test file apart from some fake test in rabbitmq. The cleanest way to fix it is refactoring the helper as you can see in this branch https://github.com/elastic/beats/pull/22573 with the solution applied as example.

ycombinator replied:
Understood, the problem is that you need to pass in the Metricset and that's not easy to construct in tests. 

I took a look at the refactoring PR (#22573) and it LGTM.

ycombinator reviewed (commented):
Left a few comments but this is starting to get pretty close!

sayden commented:
@chrisronline, @ycombinator discovered some breaking changes that have required a couple of changes in the fields of the JSON event: 

- `elasticsearch.cluster.stats.nodes.count.*` and its subkeys are now in `elasticsearch.cluster.stats.nodes.stats.*`.
- Some fields that were deleted have been restored, like `elasticsearch.cluster.stats.nodes.total`

Sorry for the change!

This metricset has now a LOT of new fields that required to be mapped. I talked with Shaunak to move the removal of fields forward but I think that in this case they are waaay too many. I'm thinking in removing the following fields if the UI doesn't get affected:

- `elasticsearch.cluster.stats.nodes.fs`
- `elasticsearch.cluster.stats.indices.segments`
- Everything under `elasticsearch.cluster.stats.indices.stack`? Or maybe just the key `ilm`

Do you have an idea if it's going to break something in the UI?

chrisronline commented:
@sayden 

> elasticsearch.cluster.stats.nodes.fs

We currently rely on:

- `node_stats.fs.total.total_in_bytes`
- `node_stats.fs.total.available_in_bytes'
- `node_stats.fs.io_stats.total.operations`
- `node_stats.fs.io_stats.total.read_operations`
- `node_stats.fs.io_stats.total.write_operations`

> elasticsearch.cluster.stats.indices.segments

We currently rely on:

- `index_stats.total.segments.memory_in_bytes`
- `index_stats.total.segments.memory_in_bytes`

sayden commented:
> 
> > elasticsearch.cluster.stats.indices.stack
> 
> What legacy field does this map to? Is this the usage/telemetry data?

@ycombinator do you know the answer of this?

ycombinator commented:
In the legacy monitoring data, i.e. documents indexed into `.monitoring-*` indices, the field was known as `stack_stats`. @sayden edited the previous comment and added a reference to the code where it's being generated by the `elasticsearch/cluster_stats` metricset (with `xpack.enabled: true`).

AFAICT, this field is expected by Telemetry code in Kibana. Searching for `stack_stats` in the Kibana repo yields several results that seem related to Telemetry: https://github.com/elastic/kibana/search?q=stack_stats. @chrisronline could you confirm that Telemetry code in Kibana looks at the `stack_stats` fields from `type: cluster_stats` documents in `.monitoring-es-*` indices? If so, then we will need to preserve this information as part of the changes in this PR, and add a field alias for it so Telemetry code continues to work as before. BTW, @chrisronline, has Telemetry code in Kibana been updated to look at `.monitoring-*` and `metricbeat-*` indices, like Stack Monitoring code?

sayden commented:
Okay I have a final version now. I had to remove 2 fields that could be affected according to this mapping https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/resources/monitoring-es.json#L503
* `cluster_state.nodes`
* `cluster_state.shards`

The reason is that an alias cannot point to an object type. According to docs _"The target must be a concrete field, and not an object or another field alias."_ 

Take a final look and tell me if I can do something as a workaround

chrisronline commented:
> could you confirm that Telemetry code in Kibana looks at the stack_stats fields from type: cluster_stats documents in .monitoring-es-* indices? 

Yes -> [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/telemetry_collection/get_es_stats.ts#L61) and [here]( https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/telemetry_collection/get_cluster_uuids.ts#L46)

> If so, then we will need to preserve this information as part of the changes in this PR, and add a field alias for it so Telemetry code continues to work as before. 

We don't need the fields aliased right? We just read them from source for telemetry purposes.

> BTW has Telemetry code in Kibana been updated to look at .monitoring-* and metricbeat-* indices, like Stack Monitoring code?

Not yet, but it will be done for https://github.com/elastic/kibana/issues/73864

sayden commented:
/test metricbeat

ycombinator commented:
> > If so, then we will need to preserve this information as part of the changes in this PR, and add a field alias for it so Telemetry code continues to work as before. 

> We don't need the fields aliased right? We just read them from source for telemetry purposes.

Gotcha. Yeah, if we just read them from source then we don't need the aliases. But then I imagine Telemetry code will need to be updated to also look at the new non-aliased field paths? Will this also be done as part of https://github.com/elastic/kibana/issues/73864?

sayden commented:
/test metricbeat
